### PR TITLE
[FLINK-2050] Introduces new pipelining mechanism for FlinkML

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR introduces the new pipelining mechanism for FlinkML. In order to make pipeline applicable to different input types, the algorithm logic and the state of the pipeline operator have been separated. The logic is now kept in implicit values which are automatically selected by the Scala compiler based on the input and output types of the pipeline operators and the input data.

The operator itself keeps now the model data which is trained in the fit phase. Thus, there is no longer a distinct model which is returned from the algorithm.

The pipelining allows, for example, a pipeline which scales vectors to work on the `Vector` type as well as `LabeledVector` type even though both types are not related. The only requirement is that implicit values implementing the algorithm are available. This approach is similar to the mechanism which can be found in the Breeze library.

### CODE REVIEW

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/StandardScaler.scala [hunk 1]
```diff
@@ -22,38 +22,47 @@ import breeze.linalg
...
+  *            val transformedDS = transformer.transform(trainingDS)
   *          }}}
   *
   * =Parameters=
   *
-  * - [[StandardScaler.Mean]]: The mean value of transformed data set; by default equal to 0
-  * - [[StandardScaler.Std]]: The standard deviation of the transformed data set; by default
+  * - [[Mean]]: The mean value of transformed data set; by default equal to 0
+  * - [[Std]]: The standard deviation of the transformed data set; by default
   * equal to 1
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/StandardScaler.scala]: Why use just the top-level type here, but the fully qualified one in the ALS docstring?

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/StandardScaler.scala]: IntelliJ did it that way. I haven't figured out yet whether the full path is necessary or not.

**Comment by rmetzger** [comment]
Documentation :white_check_mark: 
Scaladocs :white_check_mark: 
Tests :white_check_mark: 
No code outside of flink-ml touched :white_check_mark: 
:ship: it

**Comment by thvasilo** [comment]
LGTM as well.

**Comment by tillrohrmann** [comment]
That's happening because the `Estimator`, `Transformer` and `Predictor` have a default value for the ParameterMap parameter.

**Comment by tillrohrmann** [comment]
If there are no further objections, then I'll merge this PR today.

**Comment by thvasilo** [comment]
Oh that's true,yes. LGTM then!

**Pull request closed without merging**